### PR TITLE
Implement counting of enqueue/dequeue packets and errors.

### PR DESCRIPTION
**Pull request by srajag**
vif:
    - add per-lcore statistics
    - add enqueue/dequeue statistics and errors

dropstats:
    - add per-lcore statistics
    - enqueue/dequeue drop statistics

dp-core:
    - vif_drop_pkt() needs a reason it drops a packet with
    - implement enqueue/dequeue statistics
    - implement per-vif and per lcore statistics

dpdk:
    - drop packets if .f_tx() or .f_flush() failed
    - count enqueued and dequeued packets for agent, KNI and virtio interfaces

### CODE REVIEW

File: dpdk/vr_dpdk_interface.c [hunk 1]
```diff
@@ -801,13 +808,23 @@ dpdk_if_tx(struct vr_interface *vif, struct vr_packet *pkt)
...
+        ret = tx_queue->txq_ops.f_tx(tx_queue->q_queue_h, m);
+        if (ret) {
+            vif_drop_pkt(vif, vr_dpdk_mbuf_to_pkt(m), 0, VP_DROP_ENQUEUE_FAIL);
+            return -1;
+        }
+
+        if (lcore_id == vr_dpdk.packet_lcore_id) {
+            ret = tx_queue->txq_ops.f_flush(tx_queue->q_queue_h);
+            if (ret) {
+                vif_drop_pkt(vif, vr_dpdk_mbuf_to_pkt(m), 0, VP_DROP_ENQUEUE_FAIL);
```
Comment (srajag) [inline dpdk/vr_dpdk_interface.c]: When will the flush fail? Also, how do we know the number of packets that failed?

File: dpdk/vr_dpdk_lcore.c [hunk 2]
```diff
@@ -412,9 +413,12 @@ dpdk_vroute(struct vr_interface *vif, struct rte_mbuf *pkts[VR_DPDK_MAX_BURST_SZ
                 /* convert mbuf to vr_packet */
                 pkt = vr_dpdk_packet_get(mbuf, vif);
                 p_clone = vr_pclone(pkt);
-                if (likely(p_clone != NULL))
-                    monitoring_tx_queue->txq_ops.f_tx(monitoring_tx_queue->q_queue_h,
+                if (likely(p_clone != NULL)) {
+                    ret = monitoring_tx_queue->txq_ops.f_tx(monitoring_tx_queue->q_queue_h,
                         vr_dpdk_pkt_to_mbuf(p_clone));
+                    if (ret)
```
Comment (srajag) [inline dpdk/vr_dpdk_lcore.c]: Monitoring failures shouldn'e be counted against vif being monitored.

File: dpdk/vr_dpdk_virtio.c [hunk 3]
```diff
@@ -798,12 +814,16 @@ vr_dpdk_virtio_enq_pkts_to_phys_lcore(struct vr_dpdk_queue *rx_queue,
...
+        return;
 
     RTE_LOG(DEBUG, VROUTER, "%s: enqueue %u pakets to ring %p\n",
                 __func__, npkts, vq_pring);
     nb_enq = rte_ring_sp_enqueue_burst(vq_pring, (void **) pkt_arr, npkts);
-    for ( ; nb_enq < npkts; nb_enq++)
-        vr_pfree(pkt_arr[nb_enq], VP_DROP_INTERFACE_DROP);
+
+    for ( ; nb_enq < npkts; nb_enq++) {
+        vif_drop_pkt(pkt_arr[nb_enq]->vp_if, pkt_arr[nb_enq], 0, VP_DROP_ENQUEUE_FAIL);
```
Comment (srajag) [inline dpdk/vr_dpdk_virtio.c]: Can we use a different error code for this enqueue (which is for TX from the VM) and the enqueue failure when we enqueue to a ring for packets being sent to the VM?

Comment (semihalf-zmuda-wojciech) [inline dpdk/vr_dpdk_virtio.c]: > which is for TX from the VM
Isn't it RX? The comment above the function declaration says: 'enqueue packets received on a virtio interface queue onto a ring that will be handled by the lcore'.

>and the enqueue failure when we enqueue to a ring for packets being sent to the VM?
We do it in two places: dpdk_virtio_to_vm_tx() - here packets are dropped when queue is not ready to accept packets, so I consider this as ENQUEUE_FAIL. We want to send, TX queue does not cooperate, so drop. According to that, I would not change the error code.
dpdk_virtio_to_vm_flush() - here packets are processed before sending to the VM. They were dequeued somewhere else, and if something fails during processing (no address, no place in buffer), they're dropped as DEQUEUE_FAIL, so with different error code. This is what I understood from our previous email exchange on how the queues work.

**Comment by pedro-r-marques** [comment]
Please submit review requests via review.opencontrail.org.

**Pull request closed without merging**